### PR TITLE
Refactor frame template

### PR DESCRIPTION
**Pull request by OskarDamkjaer**
Currently the component tree for our frames looked like this (example where we have a single :history frame).
`Stream` -> `:history-frame` -> `frameTemplate` (containing state about fullscreen) -> `frameEditor` (containing `monaco`) which meant that when a frame was re-run (and it's frame type changed) it's children (frame template and frame editor) would unmount => meaning `monaco` would unmount, reinitialize, we'd lose fullscreen state. This PR adds a component between the `Stream.tsx` and the `Frame`s that holds the fullscreen state and the `monaco` editor, to avoid this issue.

### CODE REVIEW

**Comment by FallingCeilingS** [comment]
I think this branch should target to `cooldown` for now?

File: src/browser/modules/Stream/FrameContainer.tsx [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    isCollapsed,
+    setExportItems
+  }
+  const FrameComponent = getFrameComponent(props.frameData)
+
+  return (
+    <StyledFrame
+      className={isFullscreen ? 'is-fullscreen' : ''}
+      data-testid="frame"
+      fullscreen={isFullscreen}
```
Comment (FallingCeilingS) [inline src/browser/modules/Stream/FrameContainer.tsx:123]:
```suggestion
      ifFullscreen={isFullscreen}
```

File: src/browser/modules/Stream/FrameContainer.tsx [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    >
+      <FrameTitlebar
+        frame={frame}
+        pinned={props.frameData.isPinned}
+        fullscreen={isFullscreen}
```
Comment (FallingCeilingS) [inline src/browser/modules/Stream/FrameContainer.tsx:128]:
```suggestion
        isFullscreen={isFullscreen}
```

File: src/browser/modules/Stream/FrameContainer.tsx [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        fullscreenToggle={toggleFullscreen}
+        collapse={isCollapsed}
```
Comment (FallingCeilingS) [inline src/browser/modules/Stream/FrameContainer.tsx:130]:
```suggestion
        isCollapse={isCollapsed}
```

File: src/browser/modules/Stream/Stream.tsx [hunk 4]
```diff
@@ -21,71 +21,15 @@
 import { connect } from 'react-redux'
```
Comment (FallingCeilingS) [inline src/browser/modules/Stream/Stream.tsx:21]: Nice to see the restructuring eliminates functions that shouldn't be this component's responsibility.

File: src/browser/modules/Stream/FrameContainer.tsx [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    setFullscreen(full => !full)
+    setCollapsed(false)
+  }
+
+  return {
+    isCollapsed,
+    isFullscreen,
+    toggleCollapse,
+    toggleFullscreen
+  }
```
Comment (FallingCeilingS) [inline src/browser/modules/Stream/FrameContainer.tsx:165]: This is a hook. I noticed there's a `src/browser/hooks` folder contains several hooks, and some hooks files inside of components files or folders. Better to consider the consistency of file structure and do some refactor.
This is not related to this PR. Just write it down so we don't forget it.

**Review by FallingCeilingS — CHANGES_REQUESTED** [review]
Great job for refactoring. I have a few comments for those changes. I'll revisit it once failing tests are fixed.

File: src/browser/modules/Frame/FrameTemplate.tsx [hunk 6]
```diff
@@ -18,25 +18,18 @@
...
   StyledFrameBody,
   StyledFrameContents,
   StyledFrameStatusbar,
   StyledFrameMainSection,
-  StyledFrameAside,
-  ContentContainer
+  StyledFrameAside
 } from './styled'
 
 type FrameTemplateProps = {
```
[the remark this answers is not in the document]
Comment (OskarDamkjaer) [reply src/browser/modules/Frame/FrameTemplate.tsx:31]: Nice catch!

File: src/browser/modules/Stream/FrameContainer.tsx (hunk 3 above)
Comment (OskarDamkjaer) [reply src/browser/modules/Stream/FrameContainer.tsx:130]: Nice! refactored all of these.

File: src/browser/modules/Frame/FrameTemplate.tsx [hunk 7]
```diff
@@ -18,25 +18,18 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
```
[the remark this answers is not in the document]
Comment (OskarDamkjaer) [reply src/browser/modules/Frame/FrameTemplate.tsx:20]: Good idea! Makes it easier to understand the component I think

File: src/browser/modules/Frame/DefaultFrame.tsx [hunk 8]
```diff
@@ -20,7 +20,11 @@
 import React from 'react'
 import FrameTemplate from './FrameTemplate'
 
-const DefaultFrame = ({ frame }: any): JSX.Element => {
+const DefaultFrame = ({
+  frame,
+  isCollapsed,
+  isFullscreen
+}: any): JSX.Element => {
```
[the remark this answers is not in the document]
Comment (OskarDamkjaer) [reply src/browser/modules/Frame/DefaultFrame.tsx:27]: I suspect this frame is actually not used. I've removed it and will see if tests still pass.

[the remark this answers is not in the document]
Comment (OskarDamkjaer) [reply src/browser/modules/Frame/DefaultFrame.tsx:27]: as well as doing some manual testing

File: e2e_tests/integration/style.spec.ts [hunk 9]
```diff
@@ -50,7 +50,7 @@ describe(':style', () => {
     cy.executeCommand(':clear')
     cy.executeCommand(':style')
     // can't trigger hover with cypress so we can't know it's visible
-    cy.get('[data-testid="exportGrassButton"]').should('exist')
+    cy.get('[data-testid="exportGraSSButton"]').should('exist')
```
[the remark this answers is not in the document]
Comment (OskarDamkjaer) [reply e2e_tests/integration/style.spec.ts:53]: It's stylized like this for the export => grass is short for Graph Style Sheet

**Comment by OskarDamkjaer** [comment]
Had to do a force push to fix a merge conflict with the snapshot testing, but I should have addressed all your comments now! :)

File: src/browser/modules/Stream/FrameContainer.tsx (hunk 1 above)
Comment (FallingCeilingS) [reply src/browser/modules/Stream/FrameContainer.tsx:121]: I guess you forgot to make changes here?

File: src/browser/modules/Frame/DefaultFrame.tsx [hunk 10]
```diff
@@ -1,39 +0,0 @@
-/*
```
Comment (FallingCeilingS) [inline src/browser/modules/Frame/DefaultFrame.tsx:1]: 👍

File: src/browser/modules/Frame/FrameBodyTemplate.tsx [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+  isFullscreen: boolean
+  isCollapsed: boolean
+}
+
+function FrameBodyTemplate({
+  contents,
+  onResize = () => {
+    /*noop*/
+  },
+  sidebar,
```
Comment (FallingCeilingS) [inline src/browser/modules/Frame/FrameBodyTemplate.tsx:49]:
```suggestion
  sideBar,
```

File: src/browser/modules/Frame/FrameBodyTemplate.tsx [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+  return (
+    <>
+      <StyledFrameBody
+        fullscreen={isFullscreen}
+        collapsed={isCollapsed}
+        removePadding={removePadding}
+        hasSlides={hasSlides}
+      >
+        {sidebar && sidebar()}
```
Comment (FallingCeilingS) [inline src/browser/modules/Frame/FrameBodyTemplate.tsx:79]:
```suggestion
        {sideBar && sideBar()}
```

**Review by FallingCeilingS — CHANGES_REQUESTED** [review]
LGTM! I've tested locally and it works well. The editor is no longer re-mounted when switching between frame bodies and among normal and fullscreen views. Great job! 💯 
However, it seems like the collapse button doesn't work anymore:
Screenshot 2021-12-14 at 10 50 44
Can you please investigate it?
Also, I have some minor comments. Since this is a large PR, I'm wondering if @eijawerner wants to look at it too?

File: src/browser/modules/Stream/FrameContainer.tsx (hunk 5 above)
Comment (eijawerner) [reply src/browser/modules/Stream/FrameContainer.tsx:163]: I disagree that we should move all hooks there, I think we should only move those that are used at multiple places to the general folder. Similar to how we only move reusable components to the root, and keep those only used locally close to where it is used.

**Comment by eijawerner** [comment]
Good catch that you found that the collapse is not working @FallingCeilingS! I tested now and there is more to it, when switching to another tab after pressing the collapse it collapses, then there is no way to get back to see the tabs again...

https://user-images.githubusercontent.com/11065688/146147249-4265bb4f-77cf-4327-9c8a-e0d6539b6920.mov

**Comment by FallingCeilingS** [comment]
> Good catch that you found that the collapse is not working @FallingCeilingS! I tested now and there is more to it, when switching to another tab after pressing the collapse it collapses, then there is no way to get back to see the tabs again...
> 
>  Screen.Recording.2021-12-15.at.08.57.43.mov

@eijawerner - nice catch! 
@OskarDamkjaer - it would be great to add relevant test cases after you fixed this problem.

File: src/browser/modules/Stream/FrameContainer.tsx (hunk 1 above)
Comment (OskarDamkjaer) [reply src/browser/modules/Stream/FrameContainer.tsx:121]: Ah missed it in my search and replace.. thanks!

Comment (OskarDamkjaer) [reply src/browser/modules/Stream/FrameContainer.tsx:121]: Got a few extra unrelated ones now as well, to make the whole code base consistent

File: src/browser/modules/Frame/FrameBodyTemplate.tsx (hunk 12 above)
Comment (OskarDamkjaer) [reply src/browser/modules/Frame/FrameBodyTemplate.tsx:79]: I'd prefer to keep it as one word, like in the rest of the codebase

**Comment by OskarDamkjaer** [comment]
I've found the cause of the error was an odd pattern used in the CypherFrame. I've refactored it to avoid duplicating it's fullscreen and collapsed state in the component, and worked around the need to pass the frame height as a prop

**Review by eijawerner — CHANGES_REQUESTED** [review]
Nice cleanup and fix of bug. 👍 However when I tested this PR I discovered an issue in Safari with zoom in fullscreen not working anymore. Please have a look at that before I am ready to approve this PR!

**Review by eijawerner — APPROVED** [review]
Great work! 🏅

**Review by FallingCeilingS — APPROVED** [review]
I've tested locally and it looks good to me! Thanks for putting efforts on refactoring. 💯 
In the future, we can probably think about running our E2E tests against different browsers we support so we can identify problems much quicker: https://docs.cypress.io/guides/guides/launching-browsers#Disabled-Barriers
